script.filecleaner/default.py

```diff
@@ -0,0 +1,707 @@
+    silent = True
+    exit_status = STATUS_SUCCESS
+
+    def __init__(self):
+        debug("{0!s} version {1!s} loaded.".format(__addon__.getAddonInfo("name").decode("utf-8"),
+                                                   __addon__.getAddonInfo("version").decode("utf-8")))
+
+    def __is_canceled(self):
+        """
+        Test if the progress dialog has been canceled by the user. If the cleaner was started as a service this will
+        always return False
+        :rtype: bool
+        :return: True if the user cancelled cleaning, False otherwise.
+        """
+        if self.silent:
+            return False
+        elif self.progress.iscanceled():
+            debug("User canceled.", xbmc.LOGWARNING)
+            self.exit_status = self.STATUS_ABORTED
+            return True
+
+    def show_progress(self):
+        """
+        Toggle the progress dialog on. Use before calling the cleaning method.
+        """
+        self.silent = False
+
+    def hide_progress(self):
+        """
+        Toggle the progress dialog off. Use before calling the cleaning method.
+        """
+        self.silent = True
+
+    def clean(self, video_type):
+        """
+        Clean all watched videos of the provided type.
+
+        :type video_type: str
+        :param video_type: The type of videos to clean (one of TVSHOWS, MOVIES, MUSIC_VIDEOS).
+        :rtype: (list, int, int)
+        :return: A list of the filenames that were cleaned, as well as the number of files cleaned and the return status.
+        """
+        cleaned_files = []
+        count = 0
+        type_translation = {self.MOVIES: translate(32626), self.MUSIC_VIDEOS: translate(32627), self.TVSHOWS: translate(32628)}
+
+        if not self.silent:
+            # Cleaning <video type>
+            self.progress.update(0, translate(32629).format(type=type_translation[video_type]), *map(translate, (32615, 32615)))
+            xbmc.sleep(1000)
+
+        if video_type == self.TVSHOWS:
+            clean_this_video_type = get_setting(clean_tv_shows)
+        elif video_type == self.MOVIES:
+            clean_this_video_type = get_setting(clean_movies)
+        elif video_type == self.MUSIC_VIDEOS:
+            clean_this_video_type = get_setting(clean_music_videos)
+        else:
+            debug("Incorrect video type specified: {0!r}".format(video_type), xbmc.LOGERROR)
+            return [], 0, self.STATUS_FAILURE
+
+        progress_percent = 0
+
+        if clean_this_video_type:
+            expired_videos = self.get_expired_videos(video_type)
+            if not self.silent:
+                amount = len(expired_videos)
+                debug("Found {0} videos that may need cleaning.".format(amount))
+                try:
+                    increment = 1.0 / amount
+                except ZeroDivisionError:
+                    self.progress.update(0, *map(translate, (32621, 32622, 32623)))  # No watched videos found
+                    xbmc.sleep(2500)
```

Comment:
Please use xbmc.Monitor.waitForAbort(), rather than sleep for delays of more than 1 sec. We have problems with addons hanging up the exit process. Please see:   http://mirrors.kodi.tv/docs/python-docs/15.x-isengard/xbmc.html#Monitor-waitForAbort